play/src/hooks/useFetchAndSubscribeGameSession.tsx

```diff
@@ -0,0 +1,79 @@
+import { useState, useEffect } from 'react';
+import { useTranslation } from 'react-i18next';
+import { isNullOrUndefined, ApiClient, IGameSession } from '@righton/networking';
+
+/**
+ * Custom hook to fetch and subscribe to game session. Follows:
+ * https://react.dev/learn/you-might-not-need-an-effect#fetching-data with determination that useEffect is ok for now
+ * @param gameSessionId
+ * @param apiClient
+ * @returns
+ */
+export default function useFetchAndSubscribeGameSession(
+  gameSessionId: string,
+  apiClient: ApiClient,
+  retry: number
+) {
+  const [gameSession, setGameSession] = useState<IGameSession>();
+  const [isLoading, setIsLoading] = useState<boolean>(true);
+  const { t } = useTranslation();
+  const [error, setError] = useState<string>('');
+
+  useEffect(() => {
+    // prevents runaway condition by ignoring updates to state after component unmounts
+    let ignore = false;
+    // if player is retrying after an error, reset state values to 'isLoading' conditions unless the api call fails again
+    if (retry > 0) {
+      setIsLoading(true);
+      setError('');
+    }
+    // if gamesessionId is null, set error and prevent api calls
+    if (isNullOrUndefined(gameSessionId)){
+      setError(`${t('error.connecting.gamesessionerror')}`);
+      setIsLoading(false);
+    }
+    else {
+      apiClient
+        .getGameSession(gameSessionId)
+        .then((fetchedGame) => {
+          if (!fetchedGame)
+            setError(`${t('error.connecting.gamesessionerror')}`);
+          try {
+            const gameSessionSubscription = apiClient.subscribeUpdateGameSession(
+              fetchedGame.id,
+              (response) => {
+                if (!response)
+                  setError(`${t('error.connecting.subscriptionerror')}`);
+                // Update the gameSession object and trigger the callback
+                if (!ignore)
+                  setGameSession((prevGame) => ({ ...prevGame, ...response }));
+              }
+            );
+            setIsLoading(false);
+            return () => {
+              ignore = true;
+              gameSessionSubscription.unsubscribe();
+            };
+          } catch (e) {
+            setIsLoading(false);
+            if (e instanceof Error)
+              setError(e.message);
+            else
+              setError(`${t('error.connecting.subscriptionerror')}`);
+          }
```

Comment:
Not an expert in promises on js, but wondering if the higher level `catch`  also captures inner error handling, might worth to just throw the error here and let the outer catch unifies the error handling

Reply:
I think the weird part here is that the .that the try/catch on fetching the gamesession and the try/catch on the subscribing gamesession are separate async functions (one using .then and one using async/await). So the nested try/catch won't actually propagate up to the enclosing one (I believe). 

This got me to dig more into the way errors are being thrown around the subscription though. Currently, there is some error catching but it doesn't throw explicit errors. Additionally, feeding in an invalid gamesession doesn't throw an error from the subscribe function.

The only thing the nested try/catch tries to do right now is catch errors from the initial subscription. The only way `play` impacts this is via the gameSession.id that it passes through (and that it receives directly from the fetch prior). I think disruptions between the fetch and subscribe in the gameSession.id will be rare and all of that logic is already being tested in `networking`. Therefore, I feel that the nested try/catch is only confusing the code here. I've replaced it with a check on the `fetchedGameSession.id` prior to the subscription as a final layer of error checking here.